File: lee/src/main/java/org/example/lee/题目/图/b_课程表.java
package org.example.lee.题目.图;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Queue;

public class b_课程表 {

    /**
     * can finish
     * 拓扑排序
     * @param numCourses    num courses
     * @param prerequisites prerequisites
     * @return boolean
     */
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] arr = new int[numCourses];
        Queue<int[]> queue = new LinkedList<>();
        for(int[] ar : prerequisites){
            arr[ar[1]]++;
            queue.add(ar);
        }
        while (!queue.isEmpty()){
            //本质上一次队列就是找一次环，然后在通过安全课程减一后重新找环 直到没环
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int[] poll = queue.poll();
                if(arr[poll[0]]==0){
                    arr[poll[1]]--;
                }else {
                    queue.add(poll);
                }
            }
            if(size == queue.size()){
                return false;
            }
        }
        return true;
    }
}

File: lee/src/main/java/org/example/lee/题目/图/b_课程表.go
package 图

import "container/list"

// slice重写版本
func canFinish(numCourses int, prerequisites [][]int) bool {
	l := make([][]int, 0, numCourses)
	arr := make([]int, numCourses)
	for _, v := range prerequisites {
		arr[v[1]]++
		l = append(l, v)
	}
	for len(l) > 0 {
		nl := make([][]int, 0, numCourses)
		for _, v := range l {
			ar := v
			if arr[ar[0]] == 0 {
				arr[ar[1]]--
			} else {
				nl = append(nl, ar)
			}
		}
		if len(l) == len(nl) {
			return false
		}
		l = nl
	}
	return true
}

// list性能很差,用slice重写一下就好
func canFinish1(numCourses int, prerequisites [][]int) bool {
	l := list.New()
	arr := make([]int, numCourses)
	for _, v := range prerequisites {
		arr[v[1]]++
		l.PushFront(v)
	}
	for l.Len() > 0 {
		length := l.Len()
		for i := 0; i < length; i++ {
			ar := l.Front().Value.([]int)
			l.Remove(l.Front())
			if arr[ar[0]] == 0 {
				arr[ar[1]]--
			} else {
				l.PushBack(ar)
			}
		}
		if length == l.Len() {
			return false
		}
	}
	return true
}
